pkg/reconciler/v1alpha1/taskrun/taskrun.go

```diff
@@ -664,6 +758,45 @@ func getExceededResourcesMessage(tr *v1alpha1.TaskRun) string {
 	return fmt.Sprintf("TaskRun pod %q exceeded available resources", tr.Name)
 }
 
+// killSidecars stops all sidecar containers running inside a taskrun's pod.
+// A container is determined to be a sidecar if it is currently running - this
+// func is only expected to be called after a TaskRun completes and so all Task
+// containers should have stopped already.
+//
+// A sidecar is killed by replacing its current container image with the nop
+// image, which in turn quickly exits. If the sidedcar defines a command then
+// it will exit with a non-zero status. When we check for TaskRun success we
+// have to check for the containers we care about - not the final Pod status.
+func (c *Reconciler) killSidecars(tr *v1alpha1.TaskRun) error {
+	pod, err := c.KubeClientSet.CoreV1().Pods(tr.Namespace).Get(tr.Status.PodName, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+
+		return err
+	}
+
+	updated := false
+	if pod.Status.Phase == corev1.PodRunning {
+		for i, s := range pod.Status.ContainerStatuses {
+			if s.State.Running != nil && pod.Spec.Containers[i].Image != *nopImage {
```

Comment:
Bug here: the order of the pod's `ContainerStatuses` field doesn't always match the order of the pod's `Containers` but we're expecting that order to be consistent in order to stop the right container.